src/playground-code-editor.ts

```diff
@@ -384,17 +406,236 @@ export class PlaygroundCodeEditor extends LitElement {
         this._applyHideAndFoldRegions();
         this._showDiagnostics();
       } else {
-        // Change event is only for user input.
         this.dispatchEvent(new Event('change'));
+        this._requestCompletionsIfNeeded(_editorInstance, changeObject);
       }
     });
     this._codemirror = cm;
   }
 
+  private _requestCompletionsIfNeeded(
+    _editorInstance: Editor,
+    changeObject: EditorChange
+  ) {
+    if (!this._currentFiletypeSupportsCompletion()) return;
+
+    const previousToken = _editorInstance.getTokenAt(changeObject.from);
+    const tokenUnderCursor = this.tokenUnderCursor.string.trim();
+    const tokenUnderCursorAsString = tokenUnderCursor.trim();
+    // To help reduce round trips to a language service or a completion provider, we
+    // are providing a flag if the completion is building on top of the earlier recommendations.
+    // If the flag is true, the completion system can just filter the already stored
+    // collection of completions again with the more precise input.
+    // On deletion events, we want to query the LS again, since we might be in a new context after
+    // removing characters from our code.
+    const isInputEvent = changeObject.origin === EditorChangeOrigin.INPUT;
+    const isRefinement =
+      (tokenUnderCursor.length > 1 || previousToken.string === '.') &&
+      isInputEvent;
+    const changeWasCodeCompletion =
+      changeObject.origin === EditorChangeOrigin.COMPLETE;
+
+    if (tokenUnderCursorAsString.length <= 0) return;
```

Comment:
That's pretty much not a case that can happen.

This is due to the fact that if the change was a completion, the completion item will still be under the cursor, since the completion item won't have a space or a line break in it.